store/copr/batch_coprocessor.go

```diff
@@ -798,15 +781,99 @@ func failpointCheckWhichPolicy(act tiflashcompute.DispatchPolicy) {
 	})
 }
 
+func filterAllStoresAccordingToTiFlashReplicaRead(allStores []uint64, aliveStoreIDs map[uint64]struct{}, aliveStoreIDsInTiDBZone map[uint64]struct{}, isTiDBZoneSet bool, policy tiflash.ReplicaRead) (storesMatchedPolicy []uint64, needsCrossZoneAccess bool) {
+	if policy == tiflash.AllReplicas || !isTiDBZoneSet {
+		return allStores, false
+	}
+	// Check whether exists available stores in TiDB zone. If so, we only need to access TiFlash stores in TiDB zone.
+	for _, id := range allStores {
+		if _, ok := aliveStoreIDsInTiDBZone[id]; ok {
+			storesMatchedPolicy = append(storesMatchedPolicy, id)
+		}
+	}
+	// If no available stores in TiDB zone, we need to access TiFlash stores in other zones.
+	if len(storesMatchedPolicy) == 0 {
+		// needsCrossZoneAccess indicates whether we need to access(directly read or remote read) TiFlash stores in other zones.
+		needsCrossZoneAccess = true
+
+		if policy == tiflash.ClosestAdaptive {
+			// If the policy is `ClosestAdaptive`, we can dispatch tasks to the TiFlash stores in other zones.
+			storesMatchedPolicy = allStores
+		} else if policy == tiflash.ClosestReplicas {
+			// If the policy is `ClosestReplicas`, we dispatch tasks to the TiFlash stores in TiDB zone and remote read from other zones.
+			for id := range aliveStoreIDsInTiDBZone {
+				storesMatchedPolicy = append(storesMatchedPolicy, id)
+			}
+		}
+	}
+	return
+}
+
+// getAliveStoresAndStoreIDs gets all alive TiFlash stores and their IDs.
+// If `careTiFlashReplicaRead` is true, it will also return the IDs of the alive TiFlash stores in TiDB zone.
+func getAliveStoresAndStoreIDs(ctx context.Context, cache *RegionCache, ttl time.Duration, store *kvStore, careTiFlashReplicaRead bool, tidbZone string) (aliveStores []*tikv.Store, aliveStoreIDsInAllZones, aliveStoreIDsInTiDBZone map[uint64]struct{}) {
+	allTiFlashStores := cache.RegionCache.GetTiFlashStores(tikv.LabelFilterNoTiFlashWriteNode)
+	aliveStores = filterAliveStores(ctx, allTiFlashStores, ttl, store)
+	if careTiFlashReplicaRead {
+		aliveStoreIDsInTiDBZone = make(map[uint64]struct{}, len(aliveStores))
+		for _, as := range aliveStores {
+			// If the `zone` label of the TiFlash store is not set, we treat it as a TiFlash store in other zones.
+			if tiflashZone, isSet := as.GetLabelValue(placement.DCLabelKey); isSet && tiflashZone == tidbZone {
+				aliveStoreIDsInTiDBZone[as.StoreID()] = struct{}{}
+			}
+		}
+	} else {
+		aliveStoreIDsInAllZones = make(map[uint64]struct{}, len(aliveStores))
+		for _, as := range aliveStores {
+			aliveStoreIDsInAllZones[as.StoreID()] = struct{}{}
+		}
+	}
+	return aliveStores, aliveStoreIDsInAllZones, aliveStoreIDsInTiDBZone
```

Comment:
Seems it's meaningless the keep both `aliveStoreIDsInAllZones` and `aliveStoreIDsInTiDBZone`, can we just merge the two into one like `aliveStoreIDs`, for `AllReplicas`, it means `aliveStoreIDsInAllZones`, and for `ClosestAdaptive/ClosestReplicas` it means `aliveStoreIDsInTiDBZone`

Reply:
Oh, please ignore this, I think `aliveStoreIDsInAllZones` is still needed.